sqlserver/tests/test_metrics.py

```diff
@@ -0,0 +1,583 @@
+from datadog_checks.sqlserver.const import (
+    AO_AG_SYNC_METRICS,
+    AO_METRICS_PRIMARY,
+    AO_METRICS_SECONDARY,
+    AO_REPLICA_FAILOVER_METRICS,
+    AO_REPLICA_SYNC_METRICS,
+    DATABASE_BACKUP_METRICS,
+    DATABASE_FILES_METRICS,
+    DATABASE_FRAGMENTATION_METRICS,
+    DATABASE_INDEX_METRICS,
+    DATABASE_MASTER_FILES,
+    DATABASE_SERVICE_CHECK_NAME,
+    DATABASE_STATS_METRICS,
+    DBM_MIGRATED_METRICS,
+    INSTANCE_METRICS,
+    INSTANCE_METRICS_DATABASE,
+    OS_SCHEDULER_METRICS,
+    SERVICE_CHECK_NAME,
+    TASK_SCHEDULER_METRICS,
+    TEMPDB_FILE_SPACE_USAGE_METRICS,
+)
+
+from .common import (
+    CHECK_NAME,
+    EXPECTED_QUERY_EXECUTOR_AO_METRICS_MEMBER_COMMON,
+    EXPECTED_QUERY_EXECUTOR_AO_METRICS_PRIMARY,
+    EXPECTED_QUERY_EXECUTOR_AO_METRICS_QUORUM_COMMON,
+    EXPECTED_QUERY_EXECUTOR_AO_METRICS_REPLICA_COMMON,
+    EXPECTED_QUERY_EXECUTOR_AO_METRICS_SECONDARY,
+    SERVER_METRICS,
+)
+from .utils import always_on, not_windows_ci
+
+INCR_FRACTION_METRICS = {'sqlserver.latches.latch_wait_time'}
+AUTODISCOVERY_DBS = ['master', 'msdb', 'datadog_test']
+
+
+@pytest.mark.integration
+@pytest.mark.usefixtures('dd_environment')
+def test_check_server_metrics(
+    aggregator,
+    dd_run_check,
+    init_config,
+    instance_docker_metrics,
+):
+    sqlserver_check = SQLServer(CHECK_NAME, init_config, [instance_docker_metrics])
+    dd_run_check(sqlserver_check)
+
+    tags = instance_docker_metrics.get('tags', [])
+
+    check_sqlserver_can_connect(aggregator, instance_docker_metrics['host'], sqlserver_check.resolved_hostname, tags)
+
+    for metric_name in SERVER_METRICS:
+        expected_tags = tags
+        aggregator.assert_metric(metric_name, tags=expected_tags, hostname=sqlserver_check.resolved_hostname, count=1)
+
+
+@pytest.mark.integration
+@pytest.mark.usefixtures('dd_environment')
+@pytest.mark.parametrize("database_autodiscovery", [True, False])
+@pytest.mark.parametrize("dbm_enabled", [True, False])
+def test_check_instance_metrics(
+    aggregator,
+    dd_run_check,
+    init_config,
+    instance_docker_metrics,
+    database_autodiscovery,
+    dbm_enabled,
+):
+    instance_docker_metrics['database_autodiscovery'] = database_autodiscovery
+    instance_docker_metrics['dbm'] = dbm_enabled
+    instance_docker_metrics['include_instance_metrics'] = True
+    if database_autodiscovery:
+        instance_docker_metrics['autodiscovery_include'] = AUTODISCOVERY_DBS
+
+    sqlserver_check = SQLServer(CHECK_NAME, init_config, [instance_docker_metrics])
+    dd_run_check(sqlserver_check)
+
+    tags = instance_docker_metrics.get('tags', [])
+
+    check_sqlserver_can_connect(
+        aggregator, instance_docker_metrics['host'], sqlserver_check.resolved_hostname, tags, database_autodiscovery
+    )
+
+    for metric_name, _, _ in INSTANCE_METRICS:
+        # TODO: we should find a better way to test these metrics
+        # remove SQL Server incremental sql fraction metrics for now
```

Comment:
yeah this one is a bit annoying.. there is a class of metrics which are pretty reliable/easy to test and then some that require multiple runs of the agent. 

I almost want individual tests for all of these cases instead of parametrizing this one and having a ton of logic in here to make everything DRY'd up. E.g.: instead of having logic like this:

```
    if not database_autodiscovery:
        for metric_name, _, _ in INSTANCE_METRICS_DATABASE:
            aggregator.assert_metric(metric_name, tags=tags, hostname=sqlserver_check.resolved_hostname, count=1)
    else:
        for db in AUTODISCOVERY_DBS:
            for metric_name, _, _ in INSTANCE_METRICS_DATABASE:
                aggregator.assert_metric(
                    metric_name,
                    tags=tags + ['database:{}'.format(db)],
                    hostname=sqlserver_check.resolved_hostname,
                    count=1,
                )
  ```
  
  have a test for autodiscovery and a test for not, which might make these easier to follow when there are failures

Reply:
we do have specific tests for `test_check_incr_fraction_metrics`. I also split the metrics test into 2, with/without autodiscovery respectively